rcal-827 Add regression test for skycell mosaic generation

ddavis-stsci opened the pull request:
<!-- If this PR closes a JIRA ticket, make sure the title starts with the JIRA issue number,
for example RCAL-1234: <Fix a bug> -->
Resolves [RCAL-827](https://jira.stsci.edu/browse/RCAL-nnnn)

<!-- If this PR closes a GitHub issue, reference it here by its number -->
Closes #

<!-- describe the changes comprising this PR here -->
This PR adds a regression test for the L3 resampling of input image onto a skycell. 

**Checklist**
- [N/A] added entry in `CHANGES.rst` under the corresponding subsection
- [x] updated relevant tests
- [N/A] updated relevant documentation
- [x] updated relevant milestone(s)
- [x] added relevant label(s)
- [ ] ran regression tests, post a link to the Jenkins job below. [How to run regression tests on a PR](https://github.com/spacetelescope/romancal/wiki/Running-Regression-Tests-Against-PR-Branches)

schlafly reviewed (commented):
Can you also add the bits needed to construct L3_mosaic_asn.json and r0099101001001001001_F158_visit_r274dp63x31y81_i2d.asdf to the regtest script?  The latter just looks like strun --disable-crds-steppars roman_hlp L3_mosaic_asn.json.  Not sure if there's anything more to L3_mosaic_asn.json than asn_from_list, the three file names, and the output name?

ddavis-stsci commented:
I'm not sure why the test needs to add asn_from_list? 
In this case the L3_mosaic_asn.json is constructed to match a skycell. If the inputs are changed test will likely fail. 
So I'm not sure what adding the asn construction to the test gains us.

schlafly commented:
I just want to capture whatever you did in order to produce the L3_mosaic_asn.json.  If it's a fixed string, that's fine too, but let's put that string in the script.

schlafly commented:
I think the new version of this has an issue where you wanted to do WFI01 exposures 0001 through 0003, but you did WFI01 through WFI03 instead?

In any case, I went ahead and added the creation of these regtest files to the regtest creation script here
https://github.com/spacetelescope/romancal/pull/1228/files#diff-ab1578cd89f53a8e1bcabc27ab2bd67d6671c93b587d0236e33b1ffb81f7b0b4R189-R193
so I'm not worried about that part anymore.

ddavis-stsci commented:
Oops, Yes I did mean exposures.
Corrected, thanks

On 5/7/24 2:27 PM, Eddie Schlafly wrote:
>
> I think the new version of this has an issue where you wanted to do 
> WFI01 exposures 0001 through 0003, but you did WFI01 through WFI03 
> instead?
>
> In any case, I went ahead and added the creation of these regtest 
> files to the regtest creation script here
> https://github.com/spacetelescope/romancal/pull/1228/files#diff-ab1578cd89f53a8e1bcabc27ab2bd67d6671c93b587d0236e33b1ffb81f7b0b4R189-R193 
> <https://urldefense.com/v3/__https://github.com/spacetelescope/romancal/pull/1228/files*diff-ab1578cd89f53a8e1bcabc27ab2bd67d6671c93b587d0236e33b1ffb81f7b0b4R189-R193__;Iw!!CrWY41Z8OgsX0i-WU-0LuAcUu2o!yWI1tP8HxATAHZUj9BUhjsoeSkwJ6LHOerrQ6zo714D4sXQBdIu2LzqiM57xjaeR9fIcGfGlz-kHLbk5D9G2BoDF$>
> so I'm not worried about that part anymore.
>
> —
> Reply to this email directly, view it on GitHub 
> <https://urldefense.com/v3/__https://github.com/spacetelescope/romancal/pull/1222*issuecomment-2099051595__;Iw!!CrWY41Z8OgsX0i-WU-0LuAcUu2o!yWI1tP8HxATAHZUj9BUhjsoeSkwJ6LHOerrQ6zo714D4sXQBdIu2LzqiM57xjaeR9fIcGfGlz-kHLbk5D2UKDU-y$>, 
> or unsubscribe 
> <https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/ALXCXWPX5GR7A6W3TY4WSLLZBEMHNAVCNFSM6AAAAABHJGKEKKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAOJZGA2TCNJZGU__;!!CrWY41Z8OgsX0i-WU-0LuAcUu2o!yWI1tP8HxATAHZUj9BUhjsoeSkwJ6LHOerrQ6zo714D4sXQBdIu2LzqiM57xjaeR9fIcGfGlz-kHLbk5DxWeTfL-$>.
> You are receiving this because you authored the thread.Message ID: 
> ***@***.***>
>

braingram commented on romancal/regtest/test_mos_pipeline.py:
```diff
@@ -128,3 +129,61 @@ def test_level3_mos_pipeline(rtdata, ignore_asdf_paths):
+        rtdata.get_data(cal_file)
+
+    input_asn = "L3_mosaic_asn.json"
+    rtdata.get_data(f"WFI/image/{input_asn}")
```
This is overwriting the association you generated above with the skycell name and explains the regression test failure.

braingram replied:
This was not the cause of the failure but is an issues and explains why the output file contains:
`r0099101001001001001_F158_prompt_visit_r274dp63x31y81`
instead of:
`r0099101001001001001_r274dp63x31y81_F158_prompt_visit`

braingram commented on romancal/regtest/test_mos_pipeline.py:
```diff
@@ -128,3 +129,59 @@ def test_level3_mos_pipeline(rtdata, ignore_asdf_paths):
+        "WFI/image/r0000101001001001001_01101_0003_WFI01_cal.asdf",
+    ]
+
+    asn = asn_from_list(
+        cal_files, product_name="r0099101001001001001_r274dp63x31y81_F158_prompt_visit"
+    )
+    asn["target"] = "r274dp63x31y81"
+    asn_name = "L3_mosaic_asn.json"
+    _, serialized = asn.dump()
+    with open(asn_name, "w") as f:
+        f.write(serialized)
+
+    # asn_cmd = f"asn_from_list --product-name='r0099101001001001001_r274dp63x31y81_F158_prompt_visit'\
+    # r0000101001001001001_01101_0001_WFI01_cal.asdf \
+    # r0000101001001001001_01101_0002_WFI01_cal.asdf \
+    # r0000101001001001001_01101_0003_WFI01_cal.asdf \
+    # -o L3_m1_asn.json"
+    # os.system(asn_cmd)  # nosec
```
```suggestion
```
This association is overwritten on line 166.

braingram commented on romancal/regtest/test_mos_pipeline.py:
```diff
@@ -128,3 +129,59 @@ def test_level3_mos_pipeline(rtdata, ignore_asdf_paths):
+    diff = compare_asdf(rtdata.output, rtdata.truth, **ignore_asdf_paths)
+    assert diff.identical, diff.report()
+
+    model = rdm.open(rtdata.output, lazy_load=False)
+
+    pipeline.log.info(
+        "DMS356 MSG: Testing the creation of a Level 3 mosaic image"
+        + passfail(model.meta.cal_step.resample == "COMPLETE")
+    )
```
Also does this need to be `DMS373`?

braingram commented on romancal/regtest/test_mos_pipeline.py:
```diff
@@ -128,3 +129,52 @@ def test_level3_mos_pipeline(rtdata, ignore_asdf_paths):
+        "WFI/image/r0000101001001001001_01101_0002_WFI01_cal.asdf",
+        "WFI/image/r0000101001001001001_01101_0003_WFI01_cal.asdf",
+    ]
+
+    asn = asn_from_list(
+        cal_files, product_name="r0099101001001001001_r274dp63x31y81_F158_prompt_visit"
+    )
+    asn["target"] = "r274dp63x31y81"
+    asn_name = "L3_mosaic_asn.json"
+    _, serialized = asn.dump()
+    with open(asn_name, "w") as f:
+        f.write(serialized)
```
```suggestion
```
The association is being saved to `L3_mosaic_asn.json` which is being overwritten on line 158.
